### ISSUE TITLE
Rewrite AWS CodeDeploy Agent.

### ISSUE THREAD

**Issue by jdrago999** [aws/aws-codedeploy-agent#18]
**TL;DR** - this thing is a steaming pile.

Allow me to count the ways:

  1. Bizarre inclusion of gems inside of `vendor/*` breaks convention with basically every other gem out there.
  1. Tests do not run.
  1. No continuous integration setup.  TravisCI or similar would be a great help here.
  1. No code climate setup.
  1. Indecipherable jumps of logic to simply poll the codedeploy service and execute new deployments.
  1. No documentation.
    - log files
    - config files
    - environment variables
    - configurable options (e.g. proxy, error handling, log levels, etc)

In addition, this "issues" board is a ghost town.  Unresolved issues have lingered far too long without comment.

In need of a *working* CodeDeploy agent with actual tests, documentation and support, I propose the following:

  1. Scrap what's here.  There's nothing documented and therefore no public interface to continue to support.
  1. Document the new interface and behavior.
  1. Write tests which would prove the new interface and behavior.
  1. Write code to make those tests pass.
  1. Release the new agent as a gem.
  1. ???
  1. PROFIT!!!

Who's with me?

**Comment by cpbugbash** [comment]
jdrago999@ - We appreciate the feedback and agree that we can improve the agent repository. We want to dedicate more time to this, and here’s what we had at the top of our list:
 
1) Update this repository with the latest released version of the CodeDeploy agent.
2) Make this repository buildable and testable, and add the right documentation.
3) Make this repository the source for future agent releases.
 
We have already started work on 1 and you should see the latest version of the agent up soon. We will follow that up with 2 and 3. We would then have the ground work done to go over some of your other suggestions such as setting code coverage tools.

**Comment by jdrago999** [comment]
Thank you -

This would be a huge help.  Considering the fact that CodeDeploy is being used to manage and deploy significant investments in time and effort, at least some of that could be applied to this agent.

**Comment by jdrago999** [comment]
> 1) Update this repository with the latest released version of the CodeDeploy agent.

@cpbugbash I'd recommend that you not only update this repo with the latest released version, but also make this github repo the source of truth for the AWS codedeploy-agent itself.  No sense in pushing periodic updates to this repo if they break rank with the changes here.

**Comment by jdrago999** [comment]
> We want to dedicate more time to this

What kind of ETA can we expect?  CodeDeploy has gone without significant update since launch over six months ago.

**Comment by nathanielks** [comment]
I'm interested in hearing some more as well. I'm trying to follow the instructions [on this page](http://docs.aws.amazon.com/codedeploy/latest/userguide/how-to-run-agent.html#how-to-run-agent-install-ubuntu) and dpkg yells at me, so a functioning agent installer is of utmost importance.

**Comment by nathanielks** [comment]
Ah. I was using RVM to install ruby and the agent installer doesn't like that. Yuck.

**Comment by wmbutler** [comment]
I am so supportive of whatever it takes to make CodeDeploy good. I have literally wasted 10 hours dealing with this poorly documented and fault-prone service. Just a few examples:

permissions: syntax fails with duplicate permissions errors....er who cares if it's duplicated? If it means deploying vs. not deploying, I'm pretty sure a WARN would suffice over a fail.

ApplicationStart (really all the hooks): Why in the hell should have to make a shell script to contain the call for my regular scripts? Of course I had to discover that for myself. Then I find that I can't even call the shell script after I've copied it to the target directory. I have to leave it in the source dir. Oh, and of course I can't change permissions on it in the source directory. You guys put all this crazy effort into an appspec file when we already have shell scripts. Very very frustrating.

I could go on but I won't. Such a huge company. Can't you guys afford to pay a few people to give CodeDeploy some love?

**Comment by suryanarayanan** [comment]
Hi all,
We've been working on updating this repo to provide a better developer experience. Here's what we have done in the past couple of weeks,

1. Fixed the tests, gemspec and provided instructions to build and run the tests.
2. Cleaned up the vendor provided gems.
3. Set up CI infrastructure. Now pushes and pull requests kick off build and tests in Travis-CI.
4. Set up CodeClimate.
5. Set up code coverage.
6. Made this repo as source repo for all future CodeDeploy agent releases.

We are currently working on end to end integration tests that you can run on the repo.
Thanks.

**Comment by jdrago999** [comment]
Thank you and your team for all the hard work that you have put into these upgrades.

**Comment by suryanarayanan** [comment]
I'm closing this issue. The agent gem is being tracked in PR https://github.com/aws/aws-codedeploy-agent/pull/4  We want to have an update story in place before we can officially release a gem.
Thanks for all the valuable feedback.
